core/source/Time.mint

```diff
@@ -21,185 +36,617 @@ module Time {
+  the week within that year.
+
+    Time.calendarWeek(Time.utcDate(2016, 1, 1)) == {2016, 53}
+  */
+  fun calendarWeek (time : Time) : Tuple(Number, Number) {
     `
     (() => {
-      switch (#{what}) {
-        case 'month':
-          return DateFNS.startOfMonth(#{date})
-        case 'week':
-          return DateFNS.startOfWeek(#{date}, { weekStartsOn: 1 })
-        case 'day':
-          return DateFNS.startOfDay(#{date})
-        default:
-          return #{date}
+      let year =
+        #{time}.getUTCFullYear();
+
+      const day =
+        #{time}.getUTCDate();
+
+      const dayYear =
+        #{dayOfYear(time)};
+
+      const dayWeek =
+        #{dayOfWeekNumber(time)};
+
+      // The week number can be calculated as number of Mondays in the year up to the ordinal date.
+      // The addition by +10 consists of +7 to start the week numbering with 1
+      // instead of 0 and +3 because the first week has already started in the
+      // previous year and the first Monday is actually in week 2.
+      let weekNumber = Math.trunc((dayYear - dayWeek + 10) / 7);
+
+      if (weekNumber == 0) {
+        // Week number 0 means the date belongs to the last week of the previous year.
+        year -= 1;
+
+        // The week number depends on whether the previous year has 52 or 53 weeks
+        // which can be determined by the day of week of January 1.
+        // The year has 53 weeks if January 1 is on a Friday or the year was a leap
+        // year and January 1 is on a Saturday.
+        const janFirstDayOfWeek = (dayWeek - dayYear + 1) % 7;
+        const isLeapYear = #{Number.isLeapYear(`year`)};
+
+        if (janFirstDayOfWeek == 5 || (janFirstDayOfWeek == 6 && isLeapYear)) {
+          weekNumber = 53;
+        } else {
+          weekNumber = 52;
+        }
+      } else if (weekNumber == 53) {
+        // Week number 53 is actually week number 1 of the following year, if
+        // December 31 is on a Monday, Tuesday or Wednesday.
+        const dec31DayOfWeek = (dayWeek + 31 - day) % 7;
+
+        if (dec31DayOfWeek <= 3) {
+          weekNumber = 1;
+          year += 1;
+        }
       }
+
+      return [year, weekNumber];
     })()
     `
   }
 
-  /* Returns the end of the day / month / week of the given time. */
-  fun endOf (what : String, date : Time) : Time {
+  /*
+  Returns the day of the week of the given time (as a number from 1 to 7).
+
+    Time.dayOfWeekNumber(Time.utcDate(2018, 4, 5)) == 4
+  */
+  fun dayOfWeekNumber (time : Time) : Number {
     `
     (() => {
-      switch (#{what}) {
-        case 'month':
-          return DateFNS.endOfMonth(#{date})
-        case 'week':
-          return DateFNS.endOfWeek(#{date}, { weekStartsOn: 1 })
-        case 'day':
-          return DateFNS.endOfDay(#{date})
-        default:
-          return #{date}
-      }
+      const _ = #{time}.getUTCDay()
+      return _ === 0 ? 7 : _;
```

Comment:
Please, avoid using `_` in JS-land as it may potentially interfere with utility libraries (and it's unreadable).

```suggestion
      const dayNumber = #{time}.getUTCDay()
      return dayNumber === 0 ? 7 : dayNumber;
```